### PR TITLE
fix(solvers): solve_linear_system calls linsolve

### LINKED ISSUE #4949: solve_linear_system contains duplicate rref algorithm
```
It seems that we have two different rref (reduced row echelon form) algorithms, one in 
Matrix.rref and one in solve_linear_system in solvers.py.  I don't know too much about these 
algorithms, so I don't know if they really are the same algorithm, but in either case, I think 
solve should just be calling Matrix.rref.

This might speed things up too (examples taken from Lay's Linear Algebra, Third Edition):

In [15]: c = Matrix([[1, -2, 1, 0],[0, 2, -8, 8],[-4, 5, 9, -9]])

In [16]: c
Out[16]: 
⎡1   -2  1   0 ⎤
⎢              ⎥
⎢0   2   -8  8 ⎥
⎢              ⎥
⎣-4  5   9   -9⎦

In [17]: c.rref()
Out[17]: 
⎛⎡1  0  0  29⎤, [0, 1, 2]⎞
⎜⎢           ⎥           ⎟
⎜⎢0  1  0  16⎥           ⎟
⎜⎢           ⎥           ⎟
⎝⎣0  0  1  3 ⎦           ⎠

In [19]: solve_linear_system(c, x, y, z)
Out[19]: {x: 29, y: 16, z: 3}

In [25]: %timeit solve_linear_system(c, x, y, z)
100 loops, best of 3: 11.2 ms per loop

In [26]: %timeit c.rref()
1000 loops, best of 3: 3.59 ms per loop

In [30]: e = Matrix([[4, -1, -1, 0, 0, 0, 0, 0, 5],[-1, 4, 0, -1, 0, 0, 0, 0, 15],[-1, 0, 4, -1, -1, 0, 0, 
0, 0],[0, -1, -1, 4, 0, -1, 0, 0, 10],[0, 0, -1, 0, 4, -1, -1, 0, 0],[0, 0, 0, -1, -1, 4, 0, -1, 10],[0, 0, 
0, 0, -1, 0, 4, -1, 20],[0, 0, 0, 0, 0, -1, -1, 4, 30]])

In [32]: e.rref()
Out[32]: 
<output omitted>

In [33]: var('x1 x2 x3 x4 x5 x6 x6 x7 x8')
Out[33]: (x₁, x₂, x₃, x₄, x₅, x₆, x₆, x₇, x₈)

In [34]: solve_linear_system(e, x1, x2, x3, x4, x5, x6, x7, x8)
Out[34]: 
⎧    827      1377      886      1546      1171      1831      1967      2517⎫
⎨x₁: ───, x₂: ────, x₃: ───, x₄: ────, x₅: ────, x₆: ────, x₇: ────, x₈: 
────⎬
⎩    209      209       209      209       209       209       209       209 ⎭

In [35]: %timeit e.rref()
10 loops, best of 3: 65 ms per loop

In [36]: %timeit solve_linear_system(e, x1, x2, x3, x4, x5, x6, x7, x8)
10 loops, best of 3: 84.5 ms per loop
```

Original issue for #4949: http://code.google.com/p/sympy/issues/detail?id=1850
Original author: https://code.google.com/u/asmeurer@gmail.com/
Original owner: https://code.google.com/u/asmeurer@gmail.com/

### LINKED ISSUE #6420: solve_linear_system trouble with floats
I have a matrix like this
```
mat1=Matrix(((0,15.0,10.0,700.0),(1,1,1,100.0),(0,10.0,5.0,200.0),(-5.0,0,0,0)))
```
and a list of unknowns:
```
unknown=[a,b,c]
```
when I do this:
```
solve_linear_system(mat1,*unknown)
```
I get "None" as my answer

However if I take out any row for example `(1,1,1,100.0)`, I would get the correct answer

It will also work if I do:
```
mat1 = mat1.applyfunc(Integer)
```
Someone else pointed out that the problem seems to have something to do with using floats.

### LINKED ISSUE #18531: apart: hangs or takes too long
Usin live.sympy.org it appears that `((x**2 + 1)**3/((x - 1)**2*(x + 1)**2*(-x**2 + 2*x + 1)*(x**2 + 2*x - 1))).apart(x=x,extension={sqrt(2)})` hangs.

### PR DESCRIPTION
**Pull request by oscarbenjamin**
This makes solve_linear_system a wrapper for linsolve. Since linsolve
uses Matrix.rref which has seen recent improvements this can fix some
bugs and performance issues related to using solve_linear_system. This
also reduces the number of duplicate rref algorithms within the
codebase (GH-4949).

One test that has conditionally existent solutions is made an XFAIL. The
test is flakey and represents a behaviour of solve that does not hold in
general.

A couple of issues (GH-6420, GH-18531) with solve_linear_system are
fixed with tests added.

#### References to other Issues or PRs

Fixes #4949 
Fixes #18531 
Fixes #6420 
Closes #2275

See also #10001 

#### Brief description of what is fixed or changed

The implementation of `solve_linear_system` is remove and replaced with a simple call to `linsolve` which in turn uses `gauss_jordan_solve` which uses `Matrix.rref`. This eliminates a redundant implementation of a linear systems solver and fixes a couple of issues with solve.

#### Other comments

Another possibility would be to use rref directly in `solve_linear_system`. I notice that `linsolve` uses `rref(simplify=True)` but removing the `simplify=True` doesn't seem to make a lot of difference. A couple of tests are XFAILED as a result of this change although some other issues are fixed.

#### Release Notes

* solvers
    * solve_linear_system is now a thin wrapper for linsolve. It is recommended to use linsolve in new code. solve_linear_system maybe deprecated or removed in future.

### CODE REVIEW

File: sympy/solvers/solvers.py [hunk 1]
```diff
@@ -2301,162 +2301,22 @@ def solve_linear_system(system, *symbols, **flags):
     {}
 
     """
-    do_simplify = flags.get('simplify', True)
+    from sympy.solvers.solveset import linsolve
+    from sympy.sets import FiniteSet
 
-    if system.rows == system.cols - 1 == len(symbols):
```
Comment (smichr) [inline sympy/solvers/solvers.py]: Perhaps this case should be moved to linsolve; there are optimizations in `inv_quick` that should probably be kept.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py]: Maybe. I think that we can make rref much faster though by using the poly domains (https://github.com/sympy/sympy/issues/18445#issuecomment-578286594)

Comment (smichr) [reply sympy/solvers/solvers.py]: Until then, if my memory serves correctly, the `inv_quick` was significantly faster. I thought I added it after there was an issue raised abotu how slow solving relatively small symbolic systems was.
```python
>>> t=time();ok=inv_quick(Matrix(((x/2,y),(z,2*x))));time()-t
0.00700068473815918
>>> t=time();ok=Matrix(((x/2,y),(z,2*x))).inv();time()-t
0.41904187202453613
```

**Comment by oscarbenjamin** [comment]
Looking at the benchmarks this makes `solve` slower in some cases (although faster in others). The speed difference seems to be due to dotprodsimp which was recently added in rref and other matrix methods. That's why this fixes #18531 but also why it is slower for some unsimplifiable polynomials. Using dotprodsimp helps enormously in some problems that would otherwise hang and/or produce unsimplifiable results. However in the cases where it doesn't help it can introduce a roughly constant factor slowdown.

Here is the example from the benchmarks (simplified):
https://github.com/sympy/sympy_benchmarks/blob/0dbe7a6b906c2ee0ef3333f1eae3b510c8656bf0/benchmarks/solve.py#L47-L50
```python
from sympy import *
import sympy

class _Polynomial(object):

    expr = None
    x = sympy.Symbol('x', real=True)
    x0 = sympy.symbols('x0', real=True)
    x1 = sympy.symbols('x1', real=True)

    def __init__(self, c=None):
        self.c = c

    def diff(self, deg):
        return self.expr.diff(self.x, deg)

    # Below are properties to be subclassed
    @property
    def expr(self):
        return sum([c*self.x**o for o, c in enumerate(self.c)])

    def eval(self, x_val, deriv=0):
        return self.expr.diff(self.x, deriv).subs({self.x: x_val})


def _make_poly(order, Poly=None):
    if Poly is None:
        Poly = _Polynomial
    c = [sympy.Symbol('c_' + str(o), real=True) for o in range(order+1)]
    return Poly(c)


def _mk_eqs(wy=3, **kwargs):
    # Equations for fitting a wy*2 - 1 degree polynomial between two points,
    # at end points derivatives are known up to order: wy - 1
    p = _make_poly(2*wy-1, **kwargs)
    y0 = [sympy.Symbol('y0_' + str(i), real=True) for i in range(wy)]
    y1 = [sympy.Symbol('y1_' + str(i), real=True) for i in range(wy)]
    eqs = []
    for i in range(wy):
        eqs.append(p.diff(i).subs({p.x: p.x0}) - y0[i])
        eqs.append(p.diff(i).subs({p.x: p.x1}) - y1[i])
    return eqs, p, (y0, y1)

eqs, p, y = _mk_eqs(2) # The benchmark uses 2 because 3 is too slow
```
Using that we get (with this PR):
```julia
In [2]: eqs, p, y = _mk_eqs(2)                                                                                                                                

In [3]: eqs                                                                                                                                                   
Out[3]: 
⎡                  2        3                           2        3                               2                               2       ⎤
⎣c₀ + c₁⋅x₀ + c₂⋅x₀  + c₃⋅x₀  - y₀ ₀, c₀ + c₁⋅x₁ + c₂⋅x₁  + c₃⋅x₁  - y₁ ₀, c₁ + 2⋅c₂⋅x₀ + 3⋅c₃⋅x₀  - y₀ ₁, c₁ + 2⋅c₂⋅x₁ + 3⋅c₃⋅x₁  - y₁ ₁⎦

In [4]: p.c                                                                                                                                                   
Out[4]: [c₀, c₁, c₂, c₃]

In [5]: %time solve(eqs, *p.c)                                                                                                                                
CPU times: user 1.42 s, sys: 6.88 ms, total: 1.43 s
Wall time: 1.43 s
```
The comparable timing on master is:
```julia
In [2]: %time solve(eqs, *p.c)                                                                                                                                
CPU times: user 932 ms, sys: 5.88 ms, total: 938 ms
Wall time: 938 ms
```
So it's a substantial (1.5x) slowdown. The bulk of the time here is spent in rref and in particular with dotprodsimp trying to simplify terms using `cancel`. We can time this with rref:
```julia
In [4]: A, b = linear_eq_to_matrix(eqs, p.c)                                                                                                                  

In [5]: Aaug = Matrix.hstack(A, b)                                                                                                                            

In [6]: Aaug                                                                                                                                                  
Out[6]: 
⎡         2      3       ⎤
⎢1  x₀  x₀     x₀    y₀ ₀⎥
⎢                        ⎥
⎢         2      3       ⎥
⎢1  x₁  x₁     x₁    y₁ ₀⎥
⎢                        ⎥
⎢                 2      ⎥
⎢0  1   2⋅x₀  3⋅x₀   y₀ ₁⎥
⎢                        ⎥
⎢                 2      ⎥
⎣0  1   2⋅x₁  3⋅x₁   y₁ ₁⎦

In [7]: %time Aaug.rref()                                                                                                                                     
CPU times: user 1.07 s, sys: 3.77 ms, total: 1.08 s
Wall time: 1.09 s
```
However if we take the `DomainMatrix` from https://github.com/sympy/sympy/issues/18445#issuecomment-578286594 then we can solve this much faster:
```julia
In [5]: M = DomainMatrix.from_rows(Aaug.tolist())                                                                                                             

In [6]: %time M.rref()                                                                                                                                        
CPU times: user 73.6 ms, sys: 898 µs, total: 74.5 ms
Wall time: 75 ms

In [7]: signsimp(M.rref()[0].to_Matrix()) == Aaug.rref()[0]                                                                                                  
Out[7]: True
```
Using the larger system that is apparently too slow to benchmark we get (with this PR):
```julia
In [13]: eqs, p, y = _mk_eqs(3)                                                                                                                               

In [14]: %time solve(eqs, *p.c)        # with this PR                                                                                                                       
CPU times: user 22.1 s, sys: 80.5 ms, total: 22.2 s
Wall time: 22.3 s
```
Then with master (4x faster):
```julia
In [3]: %time solve(eqs, *p.c)                                                                                                                                
CPU times: user 5.27 s, sys: 20.5 ms, total: 5.29 s
Wall time: 5.3 s
```
However with DomainMatrix it's much faster:
```julia
In [7]: A, b = linear_eq_to_matrix(eqs, p.c)                                                                                                                  

In [8]: Aaug = Matrix.hstack(A, b)                                                                                                                            

In [9]: M = DomainMatrix.from_rows(Aaug.tolist())                                                                                                             

In [10]: %time M.rref()                                                                                                                                       
CPU times: user 203 ms, sys: 1.73 ms, total: 204 ms
Wall time: 205 ms

In [11]: signsimp(M.rref()[0].to_Matrix()) == Aaug.rref()[0]                                                                                                  
Out[11]: True
```

I'll take a look at this to see whether keeping `inv_quick` here helps with the immediate performance slowdowns that this PR has (for some cases) and potentially add it back. Longer term though I think that DomainMatrix is something that we should use in rref to speed up linsolve etc so that might be another direction to go here.

**Comment by oscarbenjamin** [comment]
Adding back `inv_quick` helps with the `_mk_eqs(3)` case above from the benchmarks but it brings back the massive slowdown in #18531:
```julia
In [1]: f = ((x**2 + 1)**3/((x - 1)**2*(x + 1)**2*(-x**2 + 2*x + 1)*(x**2 + 2*x - 1)))

In [2]: %time apart(f, x, extension={sqrt(2)})                                                                                                                
CPU times: user 4min 4s, sys: 404 ms, total: 4min 5s
Wall time: 4min 6s
```
The cause of #18531 is precisely `inv_quick`.

I think we can solve all of the performance problems mentioned here using `DomainMatrix` but we need to get `solve_linear_system` onto the same rref code as everything else as part of that (hence this PR).

**Comment by smichr** [comment]
OK, thanks for investigating. Are you ready for this to be committed?

**Comment by smichr** [comment]
Hearing nothing, I will commit this since there is a plan in place to make further improvements.

**Comment by oscarbenjamin** [comment]
Thanks @smichr. I think that there is still more work to do on this in #18844. My current plan is to try and split some parts of that out to get them merged piece by piece.

**Pull request merged**